src/asm/random.rs

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//
+// Copyright (c) 2022 Amazon.com, Inc. or its affiliates.
+//
+// Author(s):
+//   - Ali Saidi <alisaidi@amazon.com>
+
+#[derive(Copy, Clone, Debug)]
+
+/// Implement an interface for accessing Arm v8.5 RNG instructions
+/// A empty struct is used to confirm that the system has the
+/// instructions available.
+/// Example:
+/// ```no_run
+/// use cortex_a::asm::random::ArmRng;
+/// if let Some(rng) = ArmRng::new() {
+///    let rand_num = rng.rndr();
+/// }
+/// ```
+pub struct ArmRng(());
+
+use crate::registers::ID_AA64ISAR0_EL1;
+use core::arch::asm;
+use tock_registers::interfaces::Readable;
+
+impl ArmRng {
+    /// Return a empty object that is used to gate calling
+    /// rndr and rndrss on discovery of the feature so each
+    /// call doesn't need to confirm it
+    #[inline]
+    pub fn new() -> Option<Self> {
+        #[cfg(not(target_arch = "aarch64"))]
+        return None;
+
+        #[cfg(target_arch = "aarch64")]
+        if ID_AA64ISAR0_EL1.is_set(ID_AA64ISAR0_EL1::RNDR) {
+            Some(ArmRng(()))
+        } else {
+            None
+        }
+    }
+
+    /// Return an random number from the Arm v8.5 RNG
+    /// This returns an option because the instruction can fail
+    /// (e.g. the entropy is exhausted or the RNG has failed
+    #[inline]
+    pub fn rndr(self) -> Option<u64> {
+        let mut flags: u64;
+        let mut data: u64;
+
+        #[cfg(target_arch = "aarch64")]
+        unsafe {
+            asm!(
+                "mrs {o}, s3_3_c2_c4_0",
+                "mrs {f}, nzcv",
+                o = out(reg) data,
+                f = out(reg) flags,
+                options(nomem));
```

Comment:
```suggestion
                options(nomem, nostack));
```

Should be nostack as well I think?

Reply:
I'm happy to do this, but the reason i didn't was i (maybe wrongly) read into `nomem` that it must also imply `nostack` ? 

>The asm blocks does not read or write to any memory. This allows the compiler to cache the values of modified global variables in registers across the asm block since it knows that they are not read or written to by the asm.

Since on Arm the stack is also in memory doesn't nomem imply nostack?

Reply:
You're correct https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_llvm/src/asm.rs#L242-L249 shows that they're interpreted separately and if the comment is to be believed nomem is a nop.